e2e_playwright/app_hotkeys.py

```diff
@@ -12,6 +12,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import time
+
 import streamlit as st
 
 st.text_input("text_input")
+
+# Add sleep to allow capturing the running state more reliably:
+time.sleep(1)
```

Comment:
wouldn't it make sense to use the MutationObserver approach here instedad of working with sleeps?

Reply:
Yeah, but this will first need the refactoring of the MutationObserver to use the internal state

Reply:
I changed the text so that it doesn't require this wait time. Having a more generic version of `MutationObserver` would also make sense but will require more refactoring.

Reply:
I guess I would prefer to address this refactoring directly in this PR instead of having multiple different ways how to handle the state. Then we can also validate it in conjunction with the MutationObserver. If you don't find time for this I can try to look into this later this week.

Reply:
I think the test itself doesnt need this anymore. But it would be still beneficial to refactor the MutationObserver to use these states as well and to make it more generic